internals-js/src/operations.ts

```diff
@@ -1541,13 +1385,294 @@ export class SelectionSet extends Freezable<SelectionSet> {
+  const subSelectionParentType = element.kind === 'Field' ? baseType(element.definition.type!) : element.castedType();
+  if (!isCompositeType(subSelectionParentType)) {
+    // This is a leaf, so all updates should correspond ot the same field and we just use the first.
+    return selectionOfElement(element);
+  }
+
+  const subSelectionKeyedUpdates = new MultiMap<string, SelectionUpdate>();
+  for (const update of updates) {
+    if (update instanceof AbstractSelection) {
+      if (update.selectionSet) {
+        addToKeyedUpdates(subSelectionKeyedUpdates, update.selectionSet);
+      }
+    } else {
+      addSubpathToKeyUpdates(subSelectionKeyedUpdates, subSelectionParentType, update);
+    }
+  }
+  return selectionOfElement(element, makeSelectionSet(subSelectionParentType, subSelectionKeyedUpdates, fragments));
+}
+
+function updateElement(update: SelectionUpdate): OperationElement {
+  return update instanceof AbstractSelection ? update.element : update.path[0];
+}
+
+function addSubpathToKeyUpdates(
+  keyedUpdates: MultiMap<string, SelectionUpdate>,
+  subSelectionParentType: CompositeType,
+  pathUpdate: PathBasedUpdate
+) {
+  if (pathUpdate.path.length === 1) {
+    if (!pathUpdate.selections) {
+      return;
+    }
+    addToKeyedUpdates(keyedUpdates, withUnecessaryFragmentsRemoved(subSelectionParentType, pathUpdate.selections!));
+  } else {
+    keyedUpdates.add(pathUpdate.path[1].key(), { path: pathUpdate.path.slice(1), selections: pathUpdate.selections });
+  }
+}
+
+function makeSelectionSet(parentType: CompositeType, keyedUpdates: MultiMap<string, SelectionUpdate>, fragments?: NamedFragments): SelectionSet {
+  const selections = new Map<string, Selection>();
+  for (const [key, updates] of keyedUpdates.entries()) {
+    selections.set(key, makeSelection(parentType, updates, fragments));
+  }
+  return new SelectionSet(parentType, selections, fragments);
+}
+
+/**
+ * A simple wrapper over a `SelectionSetUpdates` that allows to conveniently build a selection set, then add some more updates and build it again, etc... 
+ */
+export class MutableSelectionSet<TMemoizedValue extends { [key: string]: any } = {}> {
+  private computed: SelectionSet | undefined;
+  private _memoized: TMemoizedValue | undefined;
+
+  private constructor(
+    readonly parentType: CompositeType,
+    private readonly _updates: SelectionSetUpdates,
+    private readonly memoizer: (s: SelectionSet) => TMemoizedValue,
+  ) {
+  }
+
+  static empty(parentType: CompositeType): MutableSelectionSet {
+    return this.emptyWithMemoized(parentType, () => ({}));
+  }
+
+  static emptyWithMemoized<TMemoizedValue extends { [key: string]: any }>(
+    parentType: CompositeType,
+    memoizer: (s: SelectionSet) => TMemoizedValue,
+  ): MutableSelectionSet<TMemoizedValue> {
+    return new MutableSelectionSet( parentType, new SelectionSetUpdates(), memoizer);
+  }
+
+
+  static of(selectionSet: SelectionSet): MutableSelectionSet {
```

Comment:
Is this used?

Reply:
It's indeed not used anymore, but I'd like to keep it both for symmetry with `empty` and because it make sense API wise and has a high change of being use soon enough (and it really isn't, it's hardly a maintenance burden).